mis_builder/models/aep.py

```diff
@@ -84,101 +90,105 @@ def __init__(self, companies, currency=None):
-                self._account_ids_by_code[account_code].update(account_ids)
+    def _account_codes_to_domain(self, account_codes):
+        """Convert a comma separated list of account codes
+        (possibly with wildcards) to a domain on account.account.
+        """
+        elems = []
+        for account_code in account_codes.split(','):
+            account_code = account_code.strip()
+            if '%' in account_code:
+                elems.append([('code', '=like', account_code)])
             else:
-                # search exact codes after the loop to do less queries
-                exact_codes.add(account_code)
-        for account in account_model.\
-                search([('code', 'in', list(exact_codes)),
-                        ('company_id', 'in', self.companies.ids)]):
-            self._account_ids_by_code[account.code].add(account.id)
+                elems.append([('code', '=', account_code)])
+        return tuple(expression.OR(elems))
 
     def _parse_match_object(self, mo):
         """Split a match object corresponding to an accounting variable
 
-        Returns field, mode, [account codes], (domain expression).
+        Returns field, mode, account domain, move line domain.
         """
-        field, mode, account_codes, domain = mo.groups()
-        if not mode:
-            mode = self.MODE_VARIATION
-        elif mode == 's':
-            mode = self.MODE_END
-        if account_codes.startswith('_'):
-            account_codes = account_codes[1:]
-        else:
-            account_codes = account_codes[1:-1]
-        if account_codes.strip():
-            account_codes = [a.strip() for a in account_codes.split(',')]
-        else:
-            account_codes = [None]  # None means we want all accounts
-        domain = domain or '[]'
-        eval_context = {
+        domain_eval_context = {
             'ref': self.env.ref,
             'user': self.env.user,
             'time': time,
             'datetime': datetime,
             'dateutil': dateutil,
         }
-        domain = tuple(safe_eval(domain, eval_context))
-        return field, mode, account_codes, domain
+        field, mode, account_sel, ml_domain = mo.groups()
+        # handle some legacy modes
+        if not mode:
+            mode = self.MODE_VARIATION
+        elif mode == 's':
+            mode = self.MODE_END
+        # convert account selector to account domain
+        if account_sel.startswith('_'):
+            # legacy bal_NNN%
+            acc_domain = self._account_codes_to_domain(account_sel[1:])
+        else:
+            assert account_sel[0] == '[' and account_sel[-1] == ']'
+            inner_account_sel = account_sel[1:-1].strip()
+            if not inner_account_sel:
+                # empty selector: select all accounts
+                acc_domain = tuple()
+            elif inner_account_sel.startswith('('):
+                # account selector is a domain
```

Comment:
The domain parsing is assuming the first element of the domain is a tuple - it could be a '!', '|', or '&'

Reply:
Right! Thanks for the in depth review!

Reply:
Fixed at 0314b8451a0b12228699f0ed4f5b7f329bcec5d4